### PR TITLE
DHFPROD-4105: Add API for linking/removing artifacts from steps

### PR DESCRIPTION
**Pull request by ryanjdew**

### CODE REVIEW

File: marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/impl/hub-utils.sjs [hunk 1]
```diff
@@ -192,6 +209,22 @@ class HubUtils {
     return permissions;
   }
 
+  // this function can be used to create xquery/xpath templates that are safe from injection attacks
+  xquerySanitizer(strings, ...values) {
```
Comment (rjrudin) [inline marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/impl/hub-utils.sjs]: I may have missed it, but this seems like an ideal candidate for some nice unit tests to verify that it's doing its thing correctly, as opposed to trying to verify this in the context of a complex integration test.

File: one-ui/src/main/java/com/marklogic/hub/curation/controllers/FlowController.java [hunk 2]
```diff
@@ -102,4 +108,38 @@
         return new ResponseEntity<>(HttpStatus.OK);
     }
 
+    @RequestMapping(value = "/{flowName}/steps/{stepId}/link/{artifactType}/{artifactName}", method = RequestMethod.POST)
+    @ResponseBody
+    public ResponseEntity<?> linkArtifact(@PathVariable String flowName, @PathVariable String stepId, @PathVariable String artifactType, @PathVariable String artifactName) {
+        return linkArtifact(flowName, stepId, artifactType, artifactName, null);
+    }
+
+    @RequestMapping(value = "/{flowName}/steps/{stepId}/link/{artifactType}/{artifactName}/{artifactVersion}", method = RequestMethod.POST)
```
[the remark this answers is not in the document]
Comment (rjrudin) [reply one-ui/src/main/java/com/marklogic/hub/curation/controllers/FlowController.java]: Looking at this again, I think if there's already a pattern to toss all the inputs into the URL, then this is fine. But I feel it'd be cleaner if there was an "ArtifactLink" class with type/name/version properties and a RequestBody annotation on it.

File: one-ui/src/test/java/com/marklogic/hub/curation/controllers/FlowControllerTest.java [hunk 3]
```diff
@@ -135,6 +160,9 @@ void getFlow() throws JsonProcessingException {
             catch (Exception e){
                 logger.info("Exception is expected as the step being fetched has been deleted");
             }
+        } catch (Exception e) {
+            e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (ryanjdew) [reply one-ui/src/test/java/com/marklogic/hub/curation/controllers/FlowControllerTest.java]: Removed the catch, so the exception will be thrown.

File: marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/impl/hub-utils.sjs (hunk 1 above)
Comment (ryanjdew) [reply marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/impl/hub-utils.sjs]: Added a simple ml-unit-test for this.

File: one-ui/src/main/java/com/marklogic/hub/curation/controllers/FlowController.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ryanjdew) [reply one-ui/src/main/java/com/marklogic/hub/curation/controllers/FlowController.java]: I'm leaving the API as is. The reasoning behind that is that the DELETE method won't allow for a body describing the link to be removed and it seems odd to me to have different URL patterns for linking and unlinking

[the remark this answers is not in the document]
Comment (rjrudin) [reply one-ui/src/main/java/com/marklogic/hub/curation/controllers/FlowController.java]: That's a good point, sounds good.

**Pull request merged**